File: 102.binary-tree-level-order-traversal.js
/*
 * @lc app=leetcode id=102 lang=javascript
 *
 * [102] Binary Tree Level Order Traversal
 */
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */

// summary: the level order need to be done by queue.length
// conner case:  root == null
var levelOrder = function(root) {
	if (root == null) {
		return [];
	}

	let queue = [];
	queue.push(root);
	let res = [];
	while (queue.length != 0) {
		let level = [];
		// level num is queue.length
		let len = queue.length;
		console.log('len', len);

		for (let i = 0; i < len; i++) {
			let node = queue.shift();
			console.log('node', node);
			if (node != null) level.push(node.val);

			if (node.left) queue.push(node.left);
			if (node.right) queue.push(node.right);
		}
		console.log('queue', queue);
		res.push(level);
	}

	return res;
};

File: java/102.binary-tree-level-order-traversal.java
/*
 * @lc app=leetcode id=102 lang=java
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node. public class TreeNode { int val; TreeNode
 * left; TreeNode right; TreeNode(int x) { val = x; } }
 */
class Solution {
    public List<List<Integer>> levelOrder(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();

        if (root == null)
            return res;

        Queue<Pair<TreeNode, Integer>> queue = new LinkedList<>();
        queue.offer(new Pair<TreeNode, Integer>(root, 0));

        while (!queue.isEmpty()) {
            Pair<TreeNode, Integer> cur = queue.poll();

            TreeNode node = cur.getKey();
            Integer level = cur.getValue();

            if (level == res.size())
                res.add(new ArrayList<Integer>());
            res.get(level).add(node.val);
            if (node.left != null)
                queue.offer(new Pair<TreeNode, Integer>(node.left, level + 1));
            if (node.right != null)
                queue.offer(new Pair<TreeNode, Integer>(node.right, level + 1));
        }
        return res;

    }
}
// @lc code=end
